File: scala/secret-handshake/src/main/scala/SecretHandshake.scala
import scala.util._

object SecretHandshake {
  val Handshakes: Seq[(Int, Seq[String] => Seq[String])] =
    Seq((1 << 3, "jump" +: _),
       (1 << 2, "close your eyes" +: _),
       (1 << 1, "double blink" +: _),
       (1, "wink" +: _),
       (1 << 4, _.reverse))

  def handshake(nr: Int): Seq[String] =
    (Seq.empty[String] /: Handshakes){
      case (acc, (mask, f)) => if((nr & mask) == mask) f(acc) else acc
    }

  def handshake(binary: String): Seq[String] =
    Try(BigInt(binary, 2)) match {
      case Success(nr) => handshake(nr.toInt)
      case _ => Seq.empty
    }
}

File: haskell/secret-handshake/src/SecretHandshake.hs
{-# LANGUAGE FlexibleInstances #-}
module SecretHandshake (handshake, secretWords) where

import Numeric (readInt)
import qualified Data.Char as C
import Data.Bits ((.&.))

class ToInt a where
  toInt :: a -> Int

instance ToInt Int where
  toInt = id

instance ToInt String where
  toInt = parseInt . readBinary
    where
      readBinary = readInt 2 (`elem` "01") C.digitToInt
      parseInt [(n,"")] = n
      parseInt _ = 0

secretWords :: [(Int, [[Char]] -> [[Char]])]
secretWords =
  [ (16, reverse)
  , (1, ("wink":))
  , (2, ("double blink":))
  , (4, ("close your eyes":))
  , (8, ("jump":))
  ]

handshake :: ToInt a => a -> [[Char]]
handshake n =
  foldr (say n) [] secretWords
  where
    say code (bit, f) xs =
      if bit == (bit .&. (toInt code))
        then f xs
        else xs
